subprojects/core/src/integTest/groovy/org/gradle/api/tasks/NestedInputIntegrationTest.groovy

```diff
@@ -655,6 +658,175 @@ class NestedInputIntegrationTest extends AbstractIntegrationSpec {
         fails "myTask"
     }
 
+    def "implementation of nested property in Groovy build script is tracked"() {
+        setupTaskClassWithNestedAction()
+        buildFile << """
+            task myTask(type: TaskWithNestedAction) {
+                action = ${originalImplementation}
+            }
+        """
+
+        buildFile.makeOlder()
+
+        when:
+        run 'myTask'
+        then:
+        executedAndNotSkipped(':myTask')
+
+        when:
+        buildFile.text = """
+            task myTask(type: TaskWithNestedAction) {
+                action = ${changedImplementation}
+            }
+        """
+        run 'myTask', '--info'
+        then:
+        executedAndNotSkipped(':myTask')
+        file('build/tmp/myTask/output.txt').text == "changed"
+        output.contains "Value of input property 'action.class' has changed for task ':myTask'"
+
+        where:
+        originalImplementation                  | changedImplementation
+        '{ it.text = "hello" }'                 | '{ it.text = "changed" }'
+        wrapAction('outputFile.text = "hello"') | wrapAction('outputFile.text = "changed"')
+    }
+
+    private static String wrapAction(String body) {
+        """
+            new Action() {
+                void execute(outputFile) {
+                    ${body}
+                }
+            }
+        """
+    }
+
+    @Requires(TestPrecondition.KOTLIN_SCRIPT)
+    def "implementations in nested property in Kotlin build script is tracked"() {
```

Comment:
:o: This test covers `Action<T>` coming from Kotlin DSL scripts. What about also covering regular Kotlin lambdas?

Some task implemented in Kotlin with:
```kotlin
open class MyTask : DefaultTask() {
  @get:Nested
  lateinit var lambda: () -> String
  // snip...
}
```

And the actual lambda:
```kotlin
task<MyTask>("myTask") {
  lambda = { "some text" }
}
```

Reply:
Ah, I didn't know that there is a type for lambdas in Kotlin. Is this type available from Java? I.e. can I define a task class with a property of type "lambda" in `buildSrc/src/main/java`?

Reply:
No, this would be a Kotlin task implementation in `src/main/kotlin`.

Reply:
@eskatos I added a test for Kotlin lambdas. Is this what you had in mind?

Reply:
@wolfs 👍 💯